lib/pengine/pe_output.c

```diff
@@ -1284,6 +1284,109 @@ pe__node_attribute_html(pcmk__output_t *out, va_list args) {
     return pcmk_rc_ok;
 }
 
+PCMK__OUTPUT_ARGS("node-and-op", "pe_working_set_t *", "xmlNode *")
+int
+pe__node_and_op(pcmk__output_t *out, va_list args) {
```

Comment:
What are the differences from op-history? Maybe this is one of the things worth changing for 2.1.0

Reply:
This would be a great candidate to merge into something else.  I would really kind of like to just drop the node-and-op message and reuse the stuff from crm_mon, if we could get away with it.  I introduced this to preserve existing output.

node-and-op output looks like this:
```
[root@cluster01 ~]# crm_resource --list-operations
ping    (ocf::pacemaker:ping):   Started: ping_start_0 (node=cluster02, call=16, rc=0, last-rc-change=Mon Oct 12 14:38:13 2020, exec=2053ms): complete
ping    (ocf::pacemaker:ping):   Started: ping_monitor_10000 (node=cluster02, call=17, rc=0, last-rc-change=Mon Oct 12 14:38:15 2020, exec=2039ms): complete
Fencing (stonith:fence_xvm):     Started: Fencing_monitor_0 (node=cluster02, call=10, rc=7, last-rc-change=Mon Oct 12 14:38:03 2020, exec=3ms): complete
dummy   (ocf::pacemaker:Dummy):  Started: dummy_start_0 (node=cluster02, call=18, rc=0, last-rc-change=Mon Oct 12 14:38:19 2020, exec=6019ms): complete
dummy   (ocf::pacemaker:Dummy):  Started: dummy_monitor_60000 (node=cluster02, call=19, rc=0, last-rc-change=Mon Oct 12 14:38:26 2020, exec=6015ms): complete
ping    (ocf::pacemaker:ping):   Started: ping_start_0 (node=cluster01, call=11, rc=0, last-rc-change=Mon Oct 12 14:39:52 2020, exec=2095ms): complete
ping    (ocf::pacemaker:ping):   Started: ping_monitor_10000 (node=cluster01, call=12, rc=0, last-rc-change=Mon Oct 12 14:39:54 2020, exec=2071ms): complete
Fencing (stonith:fence_xvm):     Started: Fencing_start_0 (node=cluster01, call=13, rc=0, last-rc-change=Mon Oct 12 14:39:56 2020, exec=146ms): complete
Fencing (stonith:fence_xvm):     Started: Fencing_monitor_60000 (node=cluster01, call=14, rc=0, last-rc-change=Mon Oct 12 14:39:56 2020, exec=75ms): complete
```

And op-history output looks like this:
```
[root@cluster01 ~]# crm_mon --include=none,operations -1t
Operations:
  * Node: cluster02:
    * ping: migration-threshold=1000000:
      * (16) start: last-rc-change="Mon Oct 12 14:38:13 2020" last-run="Mon Oct 12 14:38:13 2020" exec-time="2053ms" queue-time="0ms" rc=0 (ok)
      * (17) monitor: interval="10000ms" last-rc-change="Mon Oct 12 14:38:15 2020" exec-time="2039ms" queue-time="0ms" rc=0 (ok)
    * dummy: migration-threshold=1000000:
      * (14) probe: last-rc-change="Mon Oct 12 14:38:03 2020" last-run="Mon Oct 12 14:38:03 2020" exec-time="5002ms" queue-time="0ms" rc=1 (error)
      * (18) start: last-rc-change="Mon Oct 12 14:38:19 2020" last-run="Mon Oct 12 14:38:19 2020" exec-time="6019ms" queue-time="0ms" rc=0 (ok)
      * (19) monitor: interval="60000ms" last-rc-change="Mon Oct 12 14:38:26 2020" exec-time="6015ms" queue-time="0ms" rc=0 (ok)
  * Node: cluster01:
    * ping: migration-threshold=1000000:
      * (11) start: last-rc-change="Mon Oct 12 14:39:52 2020" last-run="Mon Oct 12 14:39:52 2020" exec-time="2095ms" queue-time="0ms" rc=0 (ok)
      * (12) monitor: interval="10000ms" last-rc-change="Mon Oct 12 14:39:54 2020" exec-time="2071ms" queue-time="0ms" rc=0 (ok)
    * Fencing: migration-threshold=1000000:
      * (13) start: last-rc-change="Mon Oct 12 14:39:56 2020" last-run="Mon Oct 12 14:39:56 2020" exec-time="146ms" queue-time="0ms" rc=0 (ok)
      * (14) monitor: interval="60000ms" last-rc-change="Mon Oct 12 14:39:56 2020" exec-time="75ms" queue-time="0ms" rc=0 (ok)
    * dummy: migration-threshold=1000000:
      * (18) probe: last-rc-change="Mon Oct 12 14:39:57 2020" last-run="Mon Oct 12 14:39:57 2020" exec-time="5043ms" queue-time="0ms" rc=1 (error)
      * (19) stop: last-rc-change="Mon Oct 12 14:40:02 2020" last-run="Mon Oct 12 14:40:02 2020" exec-time="6180ms" queue-time="0ms" rc=0 (ok)
```

Where the actual op-history part is the inner-most list (the lines that start with a number in parens).

Reply:
Man, they're both ugly. :) I guess let's just keep them separate for now. We can try to make them more user-friendly in the future.

Reply:
Hah, yeah, they're not great.  I suppose we could look at trying to increase user-friendliness in general for 2.1.0, though it sounds like that is going to be a busy release anyway.  There may not be time.

I would love to see crm_mon, crm_resource, and crm_simulate use the same output functions as much as possible.

Reply:
Yep, that would be a good goal for the future